packages/plugins/snark/src/index.ts

```diff
@@ -0,0 +1,285 @@
+
+  makeRelative(filepath): string {
+    return filepath.replace(`${this.embark.config.dappPath()}/`, "");
+  }
+
+  async generateSetup(name: string, config: CircuitConfig, compiledCircuit: any): Promise<{ vkPath: string, pkPath: string; }> {
+    const zkCircuit = new zkSnark.Circuit(compiledCircuit);
+    const circuitHasChanged = await this.setupTracker.hasChanged(name, config, compiledCircuit);
+    const vkPath = this.buildFilepath(name, Snarks.Extensions.VkVerifier);
+    const pkPath = this.buildFilepath(name, Snarks.Extensions.VkProof);
+    const [vkExists, pkExists] = await Promise.all([
+      this.fs.pathExists(vkPath),
+      this.fs.pathExists(pkPath)
+    ]);
+    if (!circuitHasChanged && vkExists && pkExists) {
+      this.logger.info(`${LOG_PREFIX}Using existing circuit setup for '${name}'`);
+      return { vkPath, pkPath };
+    }
+
+    this.logger.info(`${LOG_PREFIX}Grab a coffee ☕️  this could take quite a while...`);
+    const setup = zkSnark[config.protocol].setup(zkCircuit);
+    delete setup.toxic; // must discard toxic lambda
+
+    await Promise.all([
+      this.fs.writeFile(vkPath, JSON.stringify(setup.vk_verifier), "utf8"),
+      this.fs.writeFile(pkPath, JSON.stringify(setup.vk_proof), "utf8")
+    ]);
+
+    await this.setupTracker.update(name, config, compiledCircuit);
+
+    return { vkPath, pkPath };
+  }
+
+  async getCircuit(filepath) {
+    const definition = JSON.parse(await this.fs.readFile(filepath, 'utf8'));
+    return new zkSnark.Circuit(definition);
+  }
+
+  generateVerifier(name: string): Promise<{ contractPath: string, contractName: string; }> {
+    return new Promise((resolve, reject) => {
+      const vk = this.buildFilepath(name, Snarks.Extensions.VkVerifier);
+      const contractPath = path.join(
+        this.contractsBuildDir,
+        `${name}${Snarks.Extensions.Solidity}`
+      );
+      exec(
+        `${Snarks.snarkjsBinary} generateverifier --vk ${vk} -v ${contractPath}`,
+        async error => {
+          if (error) {
+            return reject(error);
+          }
+
+          const circuitName = name.charAt(0).toUpperCase() + name.substr(1).toLowerCase();
+          const contractName = `${circuitName}Verifier`;
+          let solidityOutput = await this.fs.readFile(contractPath, 'utf8');
+          solidityOutput = solidityOutput.replace("contract Verifier", `contract ${contractName}`);
+          await this.fs.writeFile(
+            contractPath,
+            solidityOutput,
+            'utf8'
+          );
+
+          resolve({ contractPath, contractName });
+        }
+      );
+    });
+  }
+
+  async compileAndGenerateContracts(contractFiles: File[], callback) {
+    try {
+      const circuits = this.circuitSetups
+        .filter((circuit) => circuit.config.exclude !== true);
+
+      await Promise.all([
+        await this.fs.ensureDir(this.buildDir),
+        await this.fs.ensureDir(this.contractsBuildDir)
+      ]);
+
+      for (const circuit of circuits) {
```

Comment:
Again, you might have ESLint issues with `await`s in `for`

Reply:
See comment above: https://github.com/embarklabs/embark/pull/2235#discussion_r373904374